examples/post-sync/main.go

```diff
@@ -0,0 +1,352 @@
+	utilruntime.Must(v1beta1.AddToScheme(scheme))
+}
+
+// Define a new struct for the unique identifier
+type SyncID struct {
+	Name      string
+	Kind      string
+	Namespace string
+}
+
+type SyncStatusController struct {
+	client        client.Client
+	log           logr.Logger
+	statusTracker *StatusTracker
+	syncKind      string // Explicit field to indicate what kind of resource we're watching
+}
+
+func (c *SyncStatusController) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	var syncID SyncID
+	var status v1beta1.Status
+	var generation int64
+	var observedGeneration int64
+	var errs, commit string
+
+	// Use the explicit syncKind to know which type to fetch
+	switch c.syncKind {
+	case "RootSync":
+		var root v1beta1.RootSync
+		if err := c.client.Get(ctx, req.NamespacedName, &root); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+		syncID = SyncID{
+			Name:      root.Name,
+			Kind:      root.Kind,
+			Namespace: root.Namespace,
+		}
+		status = root.Status.Status
+		generation = root.Generation
+		observedGeneration = root.Status.ObservedGeneration
+		errs, commit = c.processRootSyncConditions(root.Status.Conditions)
+
+	case "RepoSync":
+		var repo v1beta1.RepoSync
+		if err := c.client.Get(ctx, req.NamespacedName, &repo); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+		syncID = SyncID{
+			Name:      repo.Name,
+			Kind:      repo.Kind,
+			Namespace: req.Namespace,
+		}
+		status = repo.Status.Status
+		generation = repo.Generation
+		observedGeneration = repo.Status.ObservedGeneration
+		errs, commit = c.processRepoSyncConditions(repo.Status.Conditions)
+
+	default:
+		c.log.Error(fmt.Errorf("unknown sync kind: %s", c.syncKind), "unrecognized syncKind")
+		return reconcile.Result{}, fmt.Errorf("unknown sync kind: %s", c.syncKind)
+	}
+
+	// Process any errors from conditions
+	if err := c.handleErrors(syncID, errs, commit, generation, observedGeneration); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Process sync status
+	if !isSuccessfulSync(status) {
+		errors, commit := extractErrorAndCommitFromStatus(status)
+		if err := c.handleErrors(syncID, errors, commit, generation, observedGeneration); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// handleErrors aggregates error messages and logs them if they haven't been logged already.
+func (c *SyncStatusController) handleErrors(rsyncID SyncID, errMessage string, commit string, generation int64, observedGeneration int64) error {
+	if errMessage == "" {
+		return nil
+	}
+	if c.statusTracker.IsLogged(rsyncID, commit, errMessage) {
+		return nil
+	}
+	c.log.Info("Sync error detected", "sync", rsyncID.Name, "namespace", rsyncID.Namespace, "kind", rsyncID.Kind, "commit", commit, "status", "failed", "generation", generation, "observedGeneration", observedGeneration, "error", errMessage)
```

Comment:
nit: suggest line breaking after each pair to make this a tad easier to read at a glance, and ensure all the pairs have a key and value.

Reply:
done